src/Features/CSharp/Portable/CodeRefactorings/ConvertIfToSwitch/CSharpConvertToSwitchCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,293 @@
+                            {
+                                case SyntaxKind.IsPatternExpression:
+                                case SyntaxKind.IsExpression:
+                                case SyntaxKind.EqualsExpression:
+                                    break;
+
+                                default:
+                                    return null;
+                            }
+
+                            var p = GetPattern(leftmost);
+                            if (p == null)
+                            {
+                                return null;
+                            }
+
+                            var condition = (leftmost.Parent.Parent as BinaryExpressionSyntax)
+                                ?.WithLeft(((BinaryExpressionSyntax)leftmost.Parent).Right);
+
+                            return new Pattern.Guarded(p, (condition ?? node.Right).WalkDownParentheses());
+                        }
+
+                    default:
+                        return null;
+                }
+            }
+
+            StatementSyntax elseBody;
+            var sectionData = new List<(List<Pattern> patterns, StatementSyntax body)>();
+            var currentIfStatement = ifStatement;
+            while (true)
+            {
+                var operands = GetLogicalOrExpressionOperands(currentIfStatement.Condition).ToImmutableArray();
+                var patterns = new List<Pattern>(operands.Length);
+                foreach (var operand in operands)
+                {
+                    var pattern = GetPattern(operand);
+                    if (pattern == null)
+                    {
+                        return;
+                    }
+
+                    patterns.Insert(0, pattern);
+                }
+
+                sectionData.Add((patterns, currentIfStatement.Statement));
+                elseBody = currentIfStatement.Else?.Statement;
+                currentIfStatement = elseBody as IfStatementSyntax;
+                if (currentIfStatement == null)
+                {
+                    break;
+                }
+            }
+
+            context.RegisterRefactoring(new MyCodeAction(c =>
+            {
+                IEnumerable<SyntaxNode> GetSwitchSectionBody(StatementSyntax node)
+                {
+                    if (!(node is BlockSyntax block && block.Statements.Count == 0))
+                    {
+                        yield return node;
+                    }
+
+                    if (semanticModel.AnalyzeControlFlow(node).EndPointIsReachable)
+                    {
+                        yield return SyntaxFactory.BreakStatement();
+                    }
+                }
+
+                CasePatternSwitchLabelSyntax GetSwitchLabel(Pattern pattern)
+                {
+                    CasePatternSwitchLabelSyntax CasePatternSwitchLabel(TypeSyntax type, VariableDesignationSyntax designation)
+                        => SyntaxFactory.CasePatternSwitchLabel(
```

Comment:
sorry, but my brain is too weak to handle a nested arrow-style local function, in a local function, in a lambda.  

I get the 'purity' of scoping things so narrowly.  However, it really just interferes with clrity of what is actually going on and what the actual *logic* is.